### PR TITLE
MapGenerator -> BlankMapGenerator

### PR DESCRIPTION
**Pull request by Krogoth100**
Remove unneeded functionality of Map Generator, such as generating anything other than a plain map.
For now, no one uses the built-in map generator as a random map generator, in any capacity per se _(except me)_, because it does not have required customizability as Lua scripts do.
Main purpose is to get rid of the GL code that causes AMD fail on launch, so it can be used as a proper blank map generator of desired size.
(Resolved issue: https://github.com/beyond-all-reason/spring/issues/719)

Blank map generator is especially useful for creation of large maps, which is very impractical to do without, just by using resource-based fixed-size maps.
It may also help in future to make a better Lua-based random map generator that can generate maps of flexible size, depending on map options.

Map gen options for start-script changed:
_new_map_x_ -> _blank_map_x_
_new_map_y_ -> _blank_map_y_
(for convenience)
_blank_map_height_ (optional) added: the height that the plain mesh will be set to

### CODE REVIEW

File: rts/Map/Generation/BlankMapGenerator.cpp [hunk 1]
```diff
@@ -57,16 +97,16 @@ void CMapGenerator::GenerateSMF(CVirtualFile* fileSMF)
 	//--- Make SMFHeader ---
 	std::strcpy(smfHeader.magic, "spring map file");
 	smfHeader.version = 1;
-	smfHeader.mapid = 0x524d4746 ^ (int)setup->mapSeed;
+	smfHeader.mapid = 0; //just a random value, could be anything at this point
```
Comment (sprunk) [inline rts/Map/Generation/BlankMapGenerator.cpp:100]:
```suggestion
	smfHeader.mapid = 0; // just an arbitrary value, could be anything at this point
```

File: rts/Map/Generation/BlankMapGenerator.cpp [hunk 2]
```diff
@@ -171,80 +201,40 @@ void CMapGenerator::GenerateMapInfo(CVirtualFile* fileMapInfo)
...
-	constexpr int32_t tileBPP = 3;
-
 	//--- Make TileFileHeader ---
 	TileFileHeader smtHeader;
 	std::strcpy(smtHeader.magic, "spring tilefile");
 	smtHeader.version = 1;
-	smtHeader.numTiles = 1; //32 * 32 * (generator->GetMapSize().x * 32) * (generator->GetMapSize().y * 32);
-	smtHeader.tileSize = tileSize;
+	smtHeader.numTiles = 1; //32 * 32 * (generator->mapSize.x * 32) * (generator->mapSize.y * 32);
+	smtHeader.tileSize = 32;
```
Comment (sprunk) [inline rts/Map/Generation/BlankMapGenerator.cpp:225]: Where does this magic value come from? If the map has no texture can't it be a 512x512 tile or 1x1 tile, does it make a difference?

Comment (Krogoth100) [reply rts/Map/Generation/BlankMapGenerator.cpp:225]: > Where does this magic value come from? If the map has no texture can't it be a 512x512 tile or 1x1 tile, does it make a difference?

I think it doesn't. This part of code I just left as it was. I think it makes sense to make it 1x1.

Comment (Krogoth100) [reply rts/Map/Generation/BlankMapGenerator.cpp:225]: > Where does this magic value come from? If the map has no texture can't it be a 512x512 tile or 1x1 tile, does it make a difference?

But maybe this is not so good idea.
For some reason tileSize has a notation that it must be 32:
https://github.com/beyond-all-reason/spring/blob/0ee88788931f9f0b195eb5f895f1092fde4211c0/rts/Map/SMF/SMFFormat.h#L56

Comment (Krogoth100) [reply rts/Map/Generation/BlankMapGenerator.cpp:225]: Also this line:
https://github.com/beyond-all-reason/spring/blob/0ee88788931f9f0b195eb5f895f1092fde4211c0/rts/Map/SMF/SMFFormat.h#L118
Yea, I guess we should not touch that for now.

**Comment by lhog** [comment]
What was the GL code, AMD failed to execute?

**Comment by Krogoth100** [comment]
> What was the GL code, AMD failed to execute?

This one:
https://github.com/beyond-all-reason/spring/blob/1c217066847aefe4e090e4fb6dc67485467faeb4/rts/Map/Generation/MapGenerator.cpp#L220
Which generated gl error few lines below.
That texture it is trying to decompress is not needed at all, because narrower purpose doesn't require it.

**Comment by Krogoth100** [comment]
@lhog Reference:
https://github.com/beyond-all-reason/spring/issues/719

**Comment by lhog** [comment]
What they do there is rather silly. Can't say I fully understand the purpose.
Anyway can you run a binary with https://github.com/beyond-all-reason/spring/commit/288a24e81abbc5ca37f9c43cd25c5acb602590ce compiled in and attach an infolog here?

**Comment by Krogoth100** [comment]
> What they do there is rather silly. Can't say I fully understand the purpose. Anyway can you run a binary with [288a24e](https://github.com/beyond-all-reason/spring/commit/288a24e81abbc5ca37f9c43cd25c5acb602590ce) compiled in and attach an infolog here?

If we fix GL side and end with that, it doesn't sit too well.
A) still useless GL operations.
B) unneeded "mapSeed" parameter, which makes wrong impression (**_and_** confusion with rngSeed now).
I am planning to clean it up further, but it needs a start and better done step by step anyway.

Compiled and run the thingy:
[infolog.txt](https://github.com/beyond-all-reason/spring/files/12034111/infolog.txt)

**Comment by lhog** [comment]
Btw, why not just include a few empty maps of standard sizes?

**Comment by sprunk** [comment]
Distribution issues? Why waste space (including clutter on map list etc) on prefabricated maps, package them etc, if you can generate some ad hoc.

**Comment by Krogoth100** [comment]
> Btw, why not just include a few empty maps of standard sizes?

I moved from this way a long time ago.

> Distribution issues? Why waste space (including clutter on map list etc) on prefabricated maps, package them etc, if you can generate some ad hoc.

^This.
Generating hundreds megabytes of wasted space and forcing Recoil to load it is unacceptable.
Notice, I am using rather large maps right now, like 38x38. (Not uber, but not small either)
A really large map may even fail to load due to this.

**Comment by Krogoth100** [comment]
> Remove unneeded functionality of Map Generator, such as generating anything other than a plain map. For now, no one uses the built-in map generator as a random map generator, in any capacity per se _(except me)_, because it does not have required customizability as Lua scripts do. Main purpose is to get rid of the GL code that causes AMD fail on launch, so it can be used as a proper blank map generator of desired size. (Resolved issue: #719)
> 
> Blank map generator is especially useful for creation of large maps, which is very impractical to do without, just by using resource-based fixed-size maps. It may also help in future to make a better Lua-based random map generator that can generate maps of flexible size, depending on map options.
> 
> Map gen options for start-script changed: _new_map_x_ -> _blank_map_x_ _new_map_y_ -> _blank_map_y_ (for convenience) _blank_map_height_ (optional) added: the height that the plain mesh will be set to

**Comment by Krogoth100** [comment]
^ Did some misslicks on my laptop : P

**Comment by Krogoth100** [comment]
I believe this is also an important argument:
> It may also help in future to make a better Lua-based random map generator that can generate maps of flexible size, depending on map options.

**Comment by lhog** [comment]
Ok, please add necessary information about the change and new usage into changelog.txt

**Comment by Krogoth100** [comment]
> Ok, please add necessary information about the change and new usage into changelog.txt

Done.
Flooded with redundant commits due to an accident, never mind that.

**Comment by gajop** [comment]
Removing this was unfortunate as it breaks SpringBoard compatibility.
Could've just added the blank map feature for those who cared imo?

**Comment by Krogoth100** [comment]
> Removing this was unfortunate as it breaks SpringBoard compatibility. Could've just added the blank map feature for those who cared imo?

Didn't know that, this is the first time I hear it.
Can you elaborate how essential this function is for SpringBoard?
Could SpringBoard be updated?
The old function was removed justifiably, so its preferred to update SpringBoard.

**Comment by sprunk** [comment]
But then it will stop working for older engine versions. I think the engine shouldn't break compat if it can be trivially avoided. How about?
```diff
- const std::string* blankMapXStr = mapOpts.try_get("blank_map_x");
+ // `new_map_x` historically produced a blank map
+ const std::string* blankMapXStr = mapOpts.try_get("blank_map_x") ?: mapOpts.try_get("new_map_x");
```
and
```diff
	// before the PR:
	// file.GetTDef(mapSeed, unsigned(0), "GAME\\MapSeed");

+	// historically, "random" map generator actually just produced blank maps
+	int legacyMapSeed;
+	file.GetTDef(legacyMapSeed, unsigned(0), "GAME\\MapSeed");
+	file.GetTDef(initBlank, bool (legacyMapSeed != 0), "GAME\\InitBlank");
-	file.GetDef(initBlank, "0", "GAME\\InitBlank");
```

**Comment by Krogoth100** [comment]
> But then it will stop working for older engine versions. I think the engine shouldn't break compat if it can be trivially avoided. How about?
> 
> ```diff
> - const std::string* blankMapXStr = mapOpts.try_get("blank_map_x");
> + // `new_map_x` historically produced a blank map
> + const std::string* blankMapXStr = mapOpts.try_get("blank_map_x") ?: mapOpts.try_get("new_map_x");
> ```
> 
> and
> 
> ```diff
> 	// before the PR:
> 	// file.GetTDef(mapSeed, unsigned(0), "GAME\\MapSeed");
> 
> +	// historically, "random" map generator actually just produced blank maps
> +	int legacyMapSeed;
> +	file.GetTDef(legacyMapSeed, unsigned(0), "GAME\\MapSeed");
> +	file.GetTDef(initBlank, bool (legacyMapSeed != 0), "GAME\\InitBlank");
> -	file.GetDef(initBlank, "0", "GAME\\InitBlank");
> ```

I don't understand.
If you have old SpringBoard - use old Spring version.
If you have new SpringBoard - use new Spring version.
That's how software usually works.

If you still think it's best to pursue compatibility - sure, proceed. I may do the code if you need me.
Please clarify which things do you need then. Do you need "new_map_x" stuff, or "MapSeed" too?

If I was a developer of SpringBoard, I would be glad to take my time and update my thing to new stuff, but I am not judging  gajop here. If you really want this, it's just a few minutes of code tickling, no problem.

Although, my personal opinion is that we should pursue code integrity. If there is no actual map generation, those things should not be there.
Desire to transform the engine and clean it up must be taken into account. Compatibility stuff must not be piled up indefinitely.

**Comment by gajop** [comment]
> Can you elaborate how essential this function is for SpringBoard?

You can't even start it without it.

> Could SpringBoard be updated? 

Sure but I feel writing this code is a waste of time.

> The old function was removed justifiably, so its preferred to update SpringBoard. **Edit:** I am saying the old function function was removed, but the new one is practically just a renamed old one, which does the same. So the update must be very easy.

Supporting both old and new engines would require that I write a lot of "version/compatibility" checking code.
The functionality is similar, but not the same. The old one didn't require specifying the height, and it would generate a green map as opposed of a black one.

PS: I don't see why the original functionality was removed. I think it shouldn't have been touched, and the new feature should be just added - they can co-exist imo.

**Comment by Krogoth100** [comment]
Ok
The issue seems to be resolved with: https://github.com/beyond-all-reason/spring/pull/1093

**Pull request merged**